src/operator/numpy/random/np_bernoulli_op.h

```diff
@@ -0,0 +1,201 @@
+                         const OpContext &ctx,
+                         const std::vector<TBlob> &inputs,
+                         const std::vector<OpReqType> &req,
+                         const std::vector<TBlob> &outputs) {
+  using namespace mshadow;
+  using namespace mxnet_op;
+  const NumpyBernoulliParam &param = nnvm::get<NumpyBernoulliParam>(attrs.parsed);
+  Stream<xpu> *s = ctx.get_stream<xpu>();
+  index_t output_len = outputs[0].Size();
+  Random<xpu, float> *prnd = ctx.requested[0].get_random<xpu, float>(s);
+  Tensor<xpu, 1, float> workspace =
+      ctx.requested[1].get_space_typed<xpu, 1, float>(Shape1(output_len + 1), s);
+  Tensor<xpu, 1, float> uniform_tensor = workspace.Slice(0, output_len);
+  Tensor<xpu, 1, float> indicator_device = workspace.Slice(output_len, output_len + 1);
+  float indicator_host = 1.0;
+  float *indicator_device_ptr = indicator_device.dptr_;
+  Kernel<set_zero, xpu>::Launch(s, 1, indicator_device_ptr);
+  prnd->SampleUniform(&uniform_tensor, 0.0, 1.0);
+  if (param.prob.has_value()) {
+    // scalar prob input
+    CHECK_LE(param.prob.value(), 1.0) << "ValueError: expect probs >= 0 && probs <= 1";
+    CHECK_GE(param.prob.value(), 0.0) << "ValueError: expect probs >= 0 && probs <= 1";
+    MSHADOW_TYPE_SWITCH_WITH_BOOL(outputs[0].type_flag_, OType, {
+      Kernel<scalar_bernoulli_kernel<OType>, xpu>::Launch(
+        s, outputs[0].Size(), param.prob.value(),
+        uniform_tensor.dptr_, outputs[0].dptr<OType>());
+    });
+  } else if (param.logit.has_value()) {
+    // scalar logit input
+    // sigmoid(x) > u  <=> x > logit(u)
+    Kernel<prob_to_logit, xpu>::Launch(s, outputs[0].Size(),
+                                         uniform_tensor.dptr_);
+    MSHADOW_TYPE_SWITCH_WITH_BOOL(outputs[0].type_flag_, OType, {
+      Kernel<scalar_bernoulli_kernel<OType>, xpu>::Launch(
+        s, outputs[0].Size(), param.logit.value(),
+        uniform_tensor.dptr_, outputs[0].dptr<OType>());
+    });
+  } else {
+    if (param.is_logit) {
+      // tensor logit input
+      Kernel<prob_to_logit, xpu>::Launch(s, outputs[0].Size(),
+                                         uniform_tensor.dptr_);
+    } else {
+      // tensor prob input
+      // sigmoid(x) > u  <=> x > logit(u)
+      MSHADOW_TYPE_SWITCH(inputs[0].type_flag_, IType, {
+        Kernel<check_legal_prob_kernel<IType>, xpu>::Launch(
+            s, inputs[0].Size(), inputs[0].dptr<IType>(), indicator_device_ptr);
+      });
+      _copy<xpu>(&indicator_host, indicator_device_ptr);
+      CHECK_GE(indicator_host, 0.0)
+          << "ValueError: expect probs >= 0 && probs <= 1";
+    }
+    mxnet::TShape new_lshape, new_oshape;
+    int ndim = FillShape(inputs[0].shape_, inputs[0].shape_, outputs[0].shape_,
+                         &new_lshape, &new_lshape, &new_oshape);
+    MSHADOW_TYPE_SWITCH(inputs[0].type_flag_, IType, {
+       MSHADOW_TYPE_SWITCH_WITH_BOOL(outputs[0].type_flag_, OType, {
```

Comment:
Bad indentation here, but I'm okay with fixing this in a future PR.

Reply:
+1 for bad indentation

Reply:
Thx for pointing out, working on it